wallet/import.go

```diff
@@ -1,13 +1,191 @@
+// wallet using the legacy pay-to-pubkey (P2PK) scheme.
+func keyScopeFromPubKey(pubKey *hdkeychain.ExtendedKey,
+	addrType *waddrmgr.AddressType) (waddrmgr.KeyScope, error) {
+
+	switch pubKeyHDVersion(binary.BigEndian.Uint32(pubKey.Version())) {
+	case mainnetBIP0044, testnetBIP0044:
+		if addrType == nil {
+			return waddrmgr.KeyScope{}, errors.New("address type " +
+				"must be specified for master public key with " +
+				"legacy version")
+		}
+
+		switch *addrType {
+		case waddrmgr.NestedWitnessPubKey:
+			return waddrmgr.KeyScopeBIP0049Plus, nil
+		case waddrmgr.WitnessPubKey:
+			return waddrmgr.KeyScopeBIP0084, nil
+		default:
+			return waddrmgr.KeyScope{}, fmt.Errorf("unsupported "+
+				"address type %v", *addrType)
+		}
+
+	case mainnetBIP0049Plus, testnetBIP0049Plus:
+		return waddrmgr.KeyScopeBIP0049Plus, nil
+
+	case mainnetBIP0084, testnetBIP0084:
+		return waddrmgr.KeyScopeBIP0084, nil
+
+	default:
+		return waddrmgr.KeyScope{}, fmt.Errorf("unknown version %x",
+			pubKey.Version())
+	}
+}
+
+// validateExtendedPubKey ensures a sane derived public key is provided.
+func (w *Wallet) validateExtendedPubKey(pubKey *hdkeychain.ExtendedKey,
+	isMasterKey bool) error {
+
+	// Private keys are not allowed.
+	if pubKey.IsPrivate() {
+		return errors.New("private keys cannot be imported")
+	}
+
+	// Public key must have a version corresponding to the current chain.
+	if !pubKey.IsForNet(w.chainParams) {
+		return fmt.Errorf("expected extended public key for current "+
+			"network %v", w.chainParams.Name)
+	}
+
+	// Verify the extended public key's depth and child index based on
+	// whether it's a master key or not.
+	if isMasterKey {
+		if pubKey.Depth() != masterPubKeyDepth {
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'")
+		}
+		if pubKey.ChildIndex() < hdkeychain.HardenedKeyStart {
+			return errors.New("invalid master key, must be hardened")
+		}
+	} else {
+		if pubKey.Depth() != pubKeyDepth {
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'/change/" +
+				"address_index")
+		}
+		if pubKey.ChildIndex() >= hdkeychain.HardenedKeyStart {
+			return errors.New("invalid pulic key, must not be " +
+				"hardened")
+		}
+	}
+
+	return nil
+}
+
+// ImportAccount imports an account backed by a master extended public key. The
+// master key fingerprint denotes the fingerprint of the root key corresponding
+// to the master public key (also known as the key with derivation path m/).
+// This may be required by some hardware wallets for proper identification and
+// signing. The address type can usually be inferred from the key's version, but
+// in the case of legacy versions (xpub, tpub), an address type must be
+// specified as we intend to not support importing BIP-44 keys into the wallet
+// using the legacy pay-to-pubkey (P2PK) scheme.
+func (w *Wallet) ImportAccount(name string, masterPubKey *hdkeychain.ExtendedKey,
```

Comment:
Is this _actually_ the master xpub? Given derivation for BIP 44 is hardened all the way down to the actual account itself...

Reply:
No it's not. I kind of conflated the meaning of master xpub here with the account xpub across all my changes. I'll update those to be account xpub instead.